staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type lastAppliedUpdater struct {
+	fieldManager Manager
+}
+
+var _ Manager = &lastAppliedUpdater{}
+
+func NewLastAppliedUpdater(fieldManager Manager) Manager {
+	return &lastAppliedUpdater{
+		fieldManager: fieldManager,
+	}
+}
+
+// Update implements Manager.
+func (f *lastAppliedUpdater) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	return f.fieldManager.Update(liveObj, newObj, managed, manager)
+}
+
+// Apply keeps the client-side apply annotation up to date with
+// server-side apply managed fields
+func (f *lastAppliedUpdater) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	err := f.updateLastApplied(newObj)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to update last applied: %v", err)
+		// return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+	}
+
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+}
+
+func (f *lastAppliedUpdater) updateLastApplied(obj runtime.Object) error {
+	var accessor, err = meta.Accessor(obj)
+	if err != nil {
+		panic(fmt.Sprintf("couldn't get accessor: %v", err))
+	}
+
+	var annotations = accessor.GetAnnotations()
+	if annotations == nil {
+		return nil
+	}
+
+	oldLastApplied, ok := annotations[corev1.LastAppliedConfigAnnotation]
+	if !ok {
+		return nil
+	}
+
+	delete(annotations, corev1.LastAppliedConfigAnnotation)
+	accessor.SetAnnotations(annotations)
```

Comment:
I don't think these two lines are needed or even good here. You should probably just update rather than delete and re-insert.

Reply:
We need to remove the last-applied annotation when we serialize the object since it will be in the annotation itself. Here's the corresponding client-side apply method:

https://github.com/kubernetes/kubernetes/blob/875d3b534e3f6f118c3bf7f7e5a6b4c6ac6fe92a/staging/src/k8s.io/kubectl/pkg/util/apply.go#L71-L74

Reply:
Ah yes, your right! Thanks :-)